### PR TITLE
Add stateful rolling window

### PR DESCRIPTION
**Pull request by pawroman**
A proof of concept really, but quite happy with it. Does not depend on #22, but would be nice to avoid overlaps.

### CODE REVIEW

File: fastats/core/windowed_pass.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    ...     else:
+    ...         state[0] += val_in - val_out
+    ...     return state[0], state
+    >>> x = np.arange(7, dtype='float')
+    >>> result = windowed_stateful_pass(x, 4, value=rolling_sum)
+    >>> result
+    array([ nan,  nan,  nan,   6.,  10.,  14.,  18.])
+    """
+    result = np.full_like(x, np.nan)
+    state = np.empty(0, dtype=result.dtype)
```
Comment (pawroman) [inline fastats/core/windowed_pass.py]: Won't work with `None` - numba is too stupid for that.

File: fastats/core/windowed_pass.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+    # must call value to get the state
+    val_in = x[win-1]
+    val_out = np.nan
+    result[win-1], state = value(x[0:win], val_in, val_out, state)
+
+    # subsequent iterations can ignore state assignment
+    # (without this, numba won't optimize out the assignment)
```
Comment (pawroman) [inline fastats/core/windowed_pass.py]: What it says, without this, `state` is re-assigned on each iteration and it's very slow.

File: fastats/core/windowed_stateful_pass.py [hunk 3]
```diff
@@ -4,8 +4,8 @@
 from fastats.core.decorator import fs
 
 
-def value(x):
-    return x[0]
+def value(x, val_in, val_out, state):
```
Comment (pawroman) [inline fastats/core/windowed_stateful_pass.py]: Moved to separate module so there's no conflicts for the `value` function in #22

File: fastats/core/single_pass.py [hunk 4]
```diff
@@ -9,7 +9,7 @@ def value(x):
 
 
 @fs
-def single_pass(x):
+def single_pass(x):  # pragma: no cover
```
Comment (pawroman) [inline fastats/core/single_pass.py]: We need to work out a way of avoiding the JIT on as-needed basis, to test with both JIT and non JIT. For now, I think we should keep these as `no cover`.

**Comment by pawroman** [comment]
I renamed the file and github stupidly regards all comments as out of date - they are not!

File: fastats/core/single_pass.py (hunk 4 above)
Comment (dwillmer) [reply fastats/core/single_pass.py]: I agree that we need a proper solution to jit-vs-nonjit, but i totally disagree with `pragma: no cover` in an `@fs` function.

I think we need to maintain python and numba tests to ensure complete coverage until we have a better solution

File: fastats/core/windowed_stateful_pass.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    :func:`windowed_pass`, but expects the
+    ``value`` function be "stateful" and have this
+    return signature:
+    ``(window_arr, val_in, val_out, state_arr)``
+
+    ``val_in`` and ``val_out`` are the values
+    coming "in" to the window and already thrown
+    "out" of the window in each iteration.
+
+    The state starts off as empty array, then each
```
Comment (dwillmer) [inline fastats/core/windowed_stateful_pass.py]: `an` empty array

File: fastats/core/windowed_stateful_pass.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    successful return from the ``value`` function
+    sets the state by returning it along the
```
Comment (dwillmer) [inline fastats/core/windowed_stateful_pass.py]: `along with`

**Comment by dwillmer** [comment]
This is awesome! I especially like the notebook, very nice!

Just a couple of typos in the comments, and remove the `# pragma: no cover` and it's good to merge.

Thanks!

**Comment by dwillmer** [comment]
Also, great to see the Appveyor check running automatically, thanks for fixing that @pawroman

File: fastats/core/single_pass.py (hunk 4 above)
Comment (pawroman) [reply fastats/core/single_pass.py]: That's OK, but the target coverage won't be attainable, making builds fail.

Comment (dwillmer) [reply fastats/core/single_pass.py]: Yea, we'll take it on a case-by-case basis, but so far it's always been possible to get the coverage using pure python functions, but it's definitely slightly fiddly and a bit annoying.

I'd still prefer 'a bit annoying' for developers over 'not covered' though 😄

**Comment by pawroman** [comment]
Addressed code review comments - spectacular failure due to `Required test coverage of 98% not reached. Total coverage: 97.90%`. Happy to hear any suggestions on how to fix this.

**Comment by pawroman** [comment]
OK figured it out. I'm now keen more than ever to modify the `@fs` decorator API and how we do function injection. That's a story for another PR though :)

**Comment by dwillmer** [comment]
Yea, I think your latest change is exactly what we want to encourage anyway - the functions we're substituting in should all be tested in pure python and in numba mode.

**Comment by pawroman** [comment]
Good to merge now - I don't think I will be able to make codacy happy about the unused args.

**Comment by dwillmer** [comment]
👍 LGTM

**Pull request merged**